# FEATURE - forward message

blackironj opened the pull request:
### 추가사항
- General Rpc 요청을 받았을 때, broadcast message 일 경우 처리하는 부분
추가
- `sendMessage`에서 broadcast message일 경우 broadcast check table에 값을
넣도록 하는부분 추가
- 주기적으로 broadcast_check_table에서 보관하고 있을 최대시간을 확인  후
갱신하도록 함.
  * `refreshBroadcastCheckTable` / `monitorBroadcastMsgTable`

 ### 기타변경
-  General Rpc에서 중복된 메시지에 대한 상태값 추가
- Node간 connection 관리는 grpc channel로 관리하므로 PingPong rpc는 더이상 사용하지 않아 protobuf에서 제거
- rpc_service에서 PingPong 객체 제거

Commit:
# REMOVE - unused rpc (PingPong)

- protobuf 수정
- PingPong rpc 삭제

Commit:
# ADD - GerneralRpc Msg status

- MsgStatus_DUPLICATED 추가

Commit:
# FEATURE - Forwarding message

 ### 추가사항
- GeneralRpc 요청을 받았을 때, broadcast message 일 경우 처리하는 부분
추가
- sendMessage에서 broadcast message일 경우 broadcast check table에 값을
넣도록 하는부분 추가
- 주기적으로 broadcast_check_table에서 보관하고 있을 최대시간을 확인 후
갱신하도록 함.
  * refreshBroadcastCheckTable / monitorBroadcastMsgTable

 ### 기타변경
- rpc_service에서 PingPong 객체 제거

morishjs commented on src/plugins/net_plugin/net_plugin.cpp:
```diff
@@ -273,7 +302,7 @@ class NetPluginImpl {
           if (node.has_value()) {
             auto stub = genStub<GruutGeneralService::Stub, GruutGeneralService>(node.value().getChannelPtr());
-            auto Status = stub->GeneralService(&context, request, &msg_status);
+            auto status = stub->GeneralService(&context, request, &msg_status);
```
1. 지금 보니까 `GeneralService` 가 네이밍이 너무 모호한 것 같아요. FindNode, OpenChannel 같은건 어떤 작업을 수행하는 RPC call인지 명확한데 GeneralService를 호출하면 뭘 하는지 코드를 찾아보지 않는 이상 불명확한 것 같습니다.
2. 지금 `findNeighbors`가 timer에 의해서 30초마다 호출되고 있는데, 다음과 같은 시나리오에서 문제가 있을 것 같습니다.
- 첫번째 GeneralService 호출, 노드 #1 채널에 데이터를 전송. 요청이 올때까지 block
- 30초가 흐른 후 두번째 호출, 노드 #1 채널에 데이터를 전송하려고 하는데 다른 thread에서 channel에 lock 걸고 블록되었기 때문에 데드락 발생

blackironj replied:
1. Signer와 Merger한테 받는 메시지의 형태 ( 포장된 형태의 메시지 )가  동일합니다. 그래서 같은 rpc로 처리하려고 했습니다. 저도 이름이 모호하다고는 생각합니다. 마땅한 이름이 생각이 나지 않습니다
2. NetPluginImpl class쪽에 queue 를 하나 만들어서 쌓아두는 형태로 처리하면 어떨까요?

morishjs replied:
2. 좀 더 구체적으로 설명해주세요

blackironj replied:
2. `GeneralService` 에서 channel에 데이터를 전송할때 문제가 생기는 거니까 NetPluginImpl 쪽에 channel로 보낼 데이터를 쌓아둘 queue 를 유지하면 안 될까요?  이후 쌓인 queue에서 전송하는 형태가 되면 되지 않을까요?

blackironj replied:
제가 잘못이해했네요. channel이 내부 channel 얘기가 아니었네요 죄송합니다.

blackironj replied:
2. 해당 channel을 그냥 connection을 확인하는용도로만 사용하고 새로운 channel을 계속 만들어서 전송하고 버리는 형태가 되면 과할까요?( 엔터프라이즈서 했던 형태) 
아니면 AsyncClient 형태로 바꿔야할거 같습니다.

morishjs replied:
https://github.com/gruut/gruut/pull/24

제가 구현한 것과 같이 Block 상태가 오래가지 않도록 deadline을 설정해서 타임아웃을 발생시키도록 하는게 좋을 것 같습니다.

blackironj replied:
이 부분 deadline 설정하는 형태로 수정하겠습니다.

morishjs commented on src/plugins/net_plugin/net_plugin.cpp:
```diff
@@ -173,6 +175,31 @@ class NetPluginImpl {
+  void refreshBroadcastCheckTable() {
+    logger::INFO("Start to refresh broadcast check table");
+    if(broadcast_check_table->empty())
+      return;
```
여기서 리턴하고 더 이상 monitorBroadcastMsgTable() 이 호출되지 않을 것 같습니다.

blackironj replied:
수정하겠습니다.

park53kr commented:
broadcast 테이블을 따로 유지하는 이유가 있나요? 
broadcast해야하는 메세지이면 예를 들어, Transaction이나 Block 메세지

```c++
sendMessage {
// ...
  if(message_type == TRANSACTION || message_type == Block) {
    for(n : all_nodes) {
      request(n)
    }
  }
}
```

blackironj commented:
@park53kr  따로 유지하는 이유는 모든 노드들에게 보내는 것이 아니기 때문에 자신이 이미 포워딩/브로드캐스팅 한 메시지이면 드랍하기 위해서  메시지의 아이디를 가지고 관리하는 테이블입니다. 이것도 이름때문에 생긴 문제같네요 ㅠ 죄송합니다

Commit:
Merge branch 'develop' into feature/forward-message

Commit:
# EDIT - message 전송시 block 이 오래갈 수 있는 문제 수정

- `GeneralService` 호출 시 Block이 오래가지 않도록 timeout 설정
- 지난 commit에서 `PingPong` 관련 삭제하지 못한 부분 삭제 (PongData
struct)

park53kr commented:
@blackironj 
구체적으로 relay해야하는 메세지가 어떤거에요?

blackironj commented:
@park53kr 일단 Block 이랑 TX 정도 있을거 같습니다.
